src/pybind/mgr/cephadm/agent.py

```diff
@@ -0,0 +1,218 @@
+
+        self.cert_tmp = tempfile.NamedTemporaryFile()
+        self.cert_tmp.write(cert.encode('utf-8'))
+        self.cert_tmp.flush()  # cert_tmp must not be gc'ed
+        cert_fname = self.cert_tmp.name
+
+        verify_tls_files(cert_fname, key_fname)
+
+        self.mgr.set_uri('https://{0}:{1}/'.format(server_addr, server_port))
+
+        cherrypy.config.update({
+            'server.socket_host': server_addr,
+            'server.socket_port': server_port,
+            'engine.autoreload.on': False,
+            'server.ssl_module': 'builtin',
+            'server.ssl_certificate': cert_fname,
+            'server.ssl_private_key': key_fname,
+        })
+        cherrypy.tree.mount(
+            Root(self.mgr), '/', {'/': {'request.dispatch': cherrypy.dispatch.MethodDispatcher()}})
+        self.mgr.log.info('Starting cherrypy engine...')
+        cherrypy.engine.start()
+        self.mgr.log.info('Cherrypy engine started.')
+        # wait for the shutdown event
+        self.cherrypy_shutdown_event.wait()
+        self.cherrypy_shutdown_event.clear()
+        cherrypy.engine.stop()
+        self.mgr.log.info('Cherrypy engine stopped.')
+
+    def shutdown(self) -> None:
+        self.mgr.log.info('Stopping cherrypy engine...')
+        self.cherrypy_shutdown_event.set()
+
+
+class RequestHandler(threading.Thread):
+    def __init__(self, mgr: "CephadmOrchestrator", host: str, ls: str = '', networks: str = '', facts: str = '', ack: str = '', keyring: str = '') -> None:
+        self.mgr = mgr
+        self.host = host
+        self.ls = ls
+        self.networks = networks
+        self.facts = facts
+        self.ack = ack
+        self.keyring = keyring
+        super(RequestHandler, self).__init__(target=self.run)
+
+    def run(self) -> None:
+        try:
+            if not self.keyring:
+                self.mgr.log.debug(
+                    f'No agent keyring provided from agent on host {self.host}. Cannot verify agent')
+                return
+            if self.host not in self.mgr.cache.agent_keys[self.host]:
+                self.mgr.log.debug(
+                    f'No agent keyring stored for host {self.host}. Cannot verify agent')
+                return
+            if self.keyring != self.mgr.cache.agent_keys[self.host]:
+                self.mgr.log.debug(
+                    f'Got wrong keyring from agent on host {self.host}.')
+                return
+
+            up_to_date = False
+            if self.ack:
+                try:
+                    int_ack = int(self.ack)
+                    if self.host in self.mgr.cache.agent_counter and int_ack == self.mgr.cache.agent_counter[self.host]:
+                        up_to_date = True
+                    elif self.host in self.mgr.cache.agent_counter:
+                        # we got old counter value with message, inform agent of new timestamp
+                        self.mgr.agent_helpers._request_agent_acks({self.host})
```

Comment:
what should happen here in the case where self.host isn't in agent_counter? seems cleaner to have a ``if self.host not in self.mgr.cache.agent_counter`` check at the top and make this block the ``else:`` to avoid duplicating the check